scripts/retrieve_ci_failures.py

```diff
@@ -0,0 +1,307 @@
+        if not any(not commit["backedoutby"] for commit in obj["commits"]):
+            no_relanding_bugs.add(bug_id)
+
+    print(f"{len(no_relanding_bugs)} cases removed because there was no relanding.")
+
+    for bug_id in no_relanding_bugs:
+        del fixed_by_commit_pushes[bug_id]
+
+    print(f"{len(fixed_by_commit_pushes)} 'fixed by commit' pushes left to analyze.")
+
+    # Skip cases where there are multiple backouts associated to the same bug ID.
+    multiple_backouts = set()
+    for bug_id, backouts in backouts_by_bug_id.items():
+        if backouts > 1:
+            if bug_id in fixed_by_commit_pushes:
+                multiple_backouts.add(bug_id)
+
+    print(
+        f"{len(multiple_backouts)} cases removed because there were multiple backouts in the same bug."
+    )
+
+    for multiple_backout in multiple_backouts:
+        del fixed_by_commit_pushes[multiple_backout]
+
+    print(f"{len(fixed_by_commit_pushes)} 'fixed by commit' pushes left to analyze.")
+
+    # Skip cases where there is no backout (and so the fix was a bustage fix).
+    no_backouts = set()
+    for bug_id in fixed_by_commit_pushes.keys():
+        if bug_id not in backouts_by_bug_id:
+            print(f"No backouts on {bug_id}")
+            no_backouts.add(bug_id)
+
+    print(
+        f"{len(no_backouts)} cases removed because there were no backouts in the bug."
+    )
+
+    for no_backout in no_backouts:
+        del fixed_by_commit_pushes[no_backout]
+
+    print(f"{len(fixed_by_commit_pushes)} 'fixed by commit' pushes left to analyze.")
+
+    # TODO: skip cases where a single push contains multiple backouts?
+
+    return fixed_by_commit_pushes
+
+
+def retrieve_logs(fixed_by_commit_pushes, upload):
+    os.makedirs("ci_failures_logs", exist_ok=True)
+
+    for push in tqdm(
+        fixed_by_commit_pushes.values(), total=len(fixed_by_commit_pushes)
+    ):
+        for failure in push["failures"]:
+            task_id = failure["task_id"]
+            retry_id = failure["retry_id"]
+
+            log_path = os.path.join("ci_failures_logs", f"{task_id}.{retry_id}.log")
+            if os.path.exists(log_path) or os.path.exists(f"{log_path}.zst"):
+                continue
+
+            if upload and utils.exists_s3(log_path):
+                continue
+
+            try:
+                utils.download_check_etag(
+                    f"https://firefox-ci-tc.services.mozilla.com/api/queue/v1/task/{task_id}/runs/{retry_id}/artifacts/public/logs/live.log",
+                    log_path,
+                )
+
+                utils.zstd_compress(log_path)
+
+                os.remove(log_path)
+
+                if upload:
+                    utils.upload_s3([log_path])
+            except requests.exceptions.HTTPError:
+                pass
+
+
+def diff_failure_vs_fix(repo, failure_commits, fix_commits):
+    try:
+        fd, idx = tempfile.mkstemp(prefix="gitidx_")
+        os.close(fd)
+
+        try:
```

Comment:
Nested try-catch blocks make code less readable. If unavoidable, I would move that part to a separate function and call it here. This way, the logic with the nested try-catch would be encapsulated within the new function.